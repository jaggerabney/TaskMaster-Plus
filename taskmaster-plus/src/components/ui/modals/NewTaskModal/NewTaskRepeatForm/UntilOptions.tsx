import { UntilOptionsType, formatDate } from "@/utils/Util";

interface UntilOptionsPropsType {
  data: UntilOptionsType;
  onChange: (data: UntilOptionsType) => void;
}

const UntilOptions: React.FC<UntilOptionsPropsType> = ({ data, onChange }) => {
  const { basis, count, until } = data;

  function inputChangeHandler(
    event:
      | React.ChangeEvent<HTMLSelectElement>
      | React.ChangeEvent<HTMLInputElement>
  ) {
    switch (event.target.id) {
      case "until-basis-dropdown":
        return onChange({
          ...data,
          basis: event.target.value
        });
      case "count-input":
        return onChange({
          ...data,
          count: Number(event.target.value)
        });
      case "until-datepicker":
        const [year, month, day] = event.target.value.split("-");

        return onChange({
          ...data,
          until: new Date(Number(year), Number(month) - 1, Number(day))
        });
    }
  }

  return (
    <>
      <div
        id="until-wrapper"
        className="flex flex-row gap-4 text-xl items-center"
      >
        <input type="checkbox" className="opacity-0" />
        <label htmlFor="until-basis-dropdown">until</label>
        <select
          id="until-basis-dropdown"
          name="until-basis-dropdown"
          value={basis}
          onChange={inputChangeHandler}
          className="border border-black rounded p-2"
        >
          <option value="COUNT">after</option>
          <option value="UNTIL">date</option>
        </select>
        {basis === "COUNT" && (
          <div
            id="count-input-wrapper"
            className="flex flex-row gap-4 text-xl items-center"
          >
            <input
              id="count-input"
              type="number"
              step={1}
              min={1}
              value={count}
              onChange={inputChangeHandler}
              className="border border-black rounded p-2 w-16"
            />
            <label htmlFor="count-input">{`day${count > 1 ? "s" : ""}`}</label>
          </div>
        )}
        {basis === "UNTIL" && (
          <div
            id="until-datepicker-wrapper"
            className="flex flex-row gap-4 text-xl items-center"
          >
            <input
              id="until-datepicker"
              type="date"
              value={formatDate(until)}
              onChange={inputChangeHandler}
              className="border border-black rounded p-2"
            />
          </div>
        )}
      </div>
    </>
  );
};

export default UntilOptions;
