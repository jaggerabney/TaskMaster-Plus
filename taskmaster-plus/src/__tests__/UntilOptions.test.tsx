import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import UntilOptions from "@/components/ui/modals/NewTaskModal/NewTaskRepeatForm/UntilOptions";
import {
  defaultRepeatFormState,
  getTomorrowsDate,
  UntilOptionsType
} from "@/utils/Util";

describe("Until options component", () => {
  const mockOnChange = jest.fn();
  const mockUntilOptions = (
    <UntilOptions data={defaultRepeatFormState.until} onChange={mockOnChange} />
  );
  let mockOnChangeCalls = 0;

  it("should render correctly", () => {
    const untilOptions = render(mockUntilOptions);

    expect(untilOptions).toMatchSnapshot();
  });

  // Basis dropdown

  it("allows user to change basis dropdown", async () => {
    render(mockUntilOptions);
    const basisDropdown = document.getElementById(
      "until-basis-dropdown"
    ) as HTMLSelectElement;

    await userEvent.selectOptions(
      basisDropdown,
      screen.getByRole("option", { name: "date" })
    );

    const onChangeData: UntilOptionsType =
      mockOnChange.mock.calls[mockOnChangeCalls++][0];
    expect(onChangeData.basis).toEqual("UNTIL");
  });

  it("only has two basis options: after and date", () => {
    render(mockUntilOptions);
    const basisDropdown = document.getElementById(
      "until-basis-dropdown"
    ) as HTMLSelectElement;
    const basisDropdownOptions = [];

    for (const child of basisDropdown.children) {
      basisDropdownOptions.push(child.textContent);
    }

    expect(["after", "date"]).toEqual(basisDropdownOptions);
  });

  it('displays "after" as the default basis value', () => {
    render(mockUntilOptions);
    const basisDropdown = document.getElementById(
      "until-basis-dropdown"
    ) as HTMLSelectElement;

    expect(basisDropdown.options[basisDropdown.selectedIndex].text).toEqual(
      "after"
    );
  });

  it('displays the COUNT input when basis is "after"', async () => {
    render(mockUntilOptions);
    const countInput = document.getElementById(
      "count-input-wrapper"
    ) as HTMLDivElement;

    expect(countInput).toBeInTheDocument();
  });

  it('displays the UNTIL datepicker when basis is "date"', async () => {
    const mockUntilOptionsDateBasis = (
      <UntilOptions
        data={{
          ...defaultRepeatFormState.until,
          basis: "UNTIL"
        }}
        onChange={mockOnChange}
      />
    );

    render(mockUntilOptionsDateBasis);
    const untilDatepicker = document.getElementById(
      "until-datepicker-wrapper"
    ) as HTMLDivElement;

    expect(untilDatepicker).toBeInTheDocument();
  });

  // COUNT input

  it("allows user to change COUNT input", () => {
    render(mockUntilOptions);
    const countInput = document.getElementById(
      "count-input"
    ) as HTMLInputElement;

    fireEvent.change(countInput, { target: { value: 4 } });

    const onChangeData: UntilOptionsType =
      mockOnChange.mock.calls[mockOnChangeCalls++][0];
    expect(onChangeData.count).toBe(4);
  });

  it("sets COUNT input to 1 by default", () => {
    render(mockUntilOptions);
    const countInput = document.getElementById(
      "count-input"
    ) as HTMLInputElement;

    expect(countInput.value).toEqual("1");
  });

  // UNTIL datepicker

  it("allows user to change UNTIL datepicker", () => {
    const mockUntilOptionsDateBasis = (
      <UntilOptions
        data={{
          ...defaultRepeatFormState.until,
          basis: "UNTIL"
        }}
        onChange={mockOnChange}
      />
    );

    render(mockUntilOptionsDateBasis);
    const untilDatepicker = document.getElementById(
      "until-datepicker"
    ) as HTMLInputElement;
    fireEvent.change(untilDatepicker, { target: { value: "2030-03-14" } });

    const onChangeData: UntilOptionsType =
      mockOnChange.mock.calls[mockOnChangeCalls++][0];
    expect(onChangeData.until).toEqual(new Date(2030, 2, 14));
  });

  it("sets UNTIL datepicker to tomorrow by default", () => {
    const mockUntilOptionsDateBasis = (
      <UntilOptions
        data={{
          ...defaultRepeatFormState.until,
          basis: "UNTIL"
        }}
        onChange={mockOnChange}
      />
    );
    const tomorrowAsString = getTomorrowsDate().toISOString().split("T")[0];

    render(mockUntilOptionsDateBasis);
    const untilDatepicker = document.getElementById(
      "until-datepicker"
    ) as HTMLInputElement;

    expect(untilDatepicker.value).toMatch(tomorrowAsString);
  });
});
