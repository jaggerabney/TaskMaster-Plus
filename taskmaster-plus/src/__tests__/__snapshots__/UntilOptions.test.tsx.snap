// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Until options component should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-row gap-4 text-xl items-center"
        id="until-wrapper"
      >
        <input
          class="opacity-0"
          type="checkbox"
        />
        <label
          for="until-basis-dropdown"
        >
          until
        </label>
        <select
          class="border border-black rounded p-2"
          id="until-basis-dropdown"
          name="until-basis-dropdown"
        >
          <option
            value="COUNT"
          >
            after
          </option>
          <option
            value="UNTIL"
          >
            date
          </option>
        </select>
        <div
          class="flex flex-row gap-4 text-xl items-center"
          id="count-input-wrapper"
        >
          <input
            class="border border-black rounded p-2 w-16"
            id="count-input"
            min="1"
            step="1"
            type="number"
            value="1"
          />
          <label
            for="count-input"
          >
            day
          </label>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-row gap-4 text-xl items-center"
      id="until-wrapper"
    >
      <input
        class="opacity-0"
        type="checkbox"
      />
      <label
        for="until-basis-dropdown"
      >
        until
      </label>
      <select
        class="border border-black rounded p-2"
        id="until-basis-dropdown"
        name="until-basis-dropdown"
      >
        <option
          value="COUNT"
        >
          after
        </option>
        <option
          value="UNTIL"
        >
          date
        </option>
      </select>
      <div
        class="flex flex-row gap-4 text-xl items-center"
        id="count-input-wrapper"
      >
        <input
          class="border border-black rounded p-2 w-16"
          id="count-input"
          min="1"
          step="1"
          type="number"
          value="1"
        />
        <label
          for="count-input"
        >
          day
        </label>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
